lesson21.md

```diff
@@ -0,0 +1,211 @@
+```
+
+Query parameters are a list of key/value pairs:
+
+- the list starts with a question mark `?`
+- each pair of key/value is separated by a `&`
+- between key and value there is a `=` sign
+
+---
+
+### Query parameter: example
+
+Whenever we search something on Google, for example, the keyword we searched is passed as a query parameter.
+
+For example:
+
+https://google.com/search?q=brontosaurus
+
+---
+
+### Query parameters and APIs
+
+Many APIs make use of query parameters to offer extra functionalities.
+
+With the Rest Countries API we can pass a `fields` key to limit the number of fields we want in the resulting objects.
+
+---
+
+### Filtering countries result
+
+- https://restcountries.eu/rest/v2/all?fields=name;capital;currencies
+
+- https://restcountries.eu/rest/v2/name/germany?fields=name;capital;currencies;languages
+
+
+---
+
+### The Rick and Morty API
+
+https://rickandmortyapi.com
+
+Base url: https://rickandmortyapi.com/api
+
+Endpoints:
+
+- https://rickandmortyapi.com/api/character
+- https://rickandmortyapi.com/api/location
+- https://rickandmortyapi.com/api/episode
+
+Let's use the one for characters.
+
+---
+
+### The Rick and Morty API
+
+Get a list of all the characters
+
+```js
+async function getAllCharacters() {
+  let response = await fetch("https://rickandmortyapi.com/api/character");
+  let result = await response.json();
+  console.log(result);
+}
+```
+
+---
+
+### The Rick and Morty API
+
+Get a single character using the id:
+
+```js
+async function getAllCharacters() {
+  let response = await fetch("https://rickandmortyapi.com/api/character/1");
+  let result = await response.json();
+  console.log(result);
+}
+```
+
+---
+
+### The Rick and Morty API
+
+```js
+// Filter characters by name:
+// https://rickandmortyapi.com/api/character/?name=rick
+
+// Filter characters by name and species: 
+// https://rickandmortyapi.com/api/character/?name=rick&species=Human
+
+---
+
+### The Open Trivia Database API
+
+```js
+async function fetchMyApi() {
+  let response = await fetch("https://opentdb.com/api.php?amount=10&category=9");
+  let result = await response.json();
+  console.log(result);
+}
+```
+
+---
+
+### Exercise 1: Rick and Morty API
+
+https://rickandmortyapi.com/documentation/#get-all-characters
+
+1. use the Rick and Morty API to get a list of all characters
+1. use the result to display some info on the page:
+    - name of the character
+    - picture
+    - species and status
+
+Look at the example on the next slide, but feel free to style as you wish.
+
+---
+
+### Exercise 1: Rick and Morty API
+
+
+![](images/lesson-21-rick-and-morty-api-basic.png) <!-- .element: style="width: 600px"-->
+
+---
+
+### Exercise 2: Rick and Morty API
+
+Use the code you've already written in exercise 1. 
+
+1. add an input field and a button to search characters by name 
+1. the list of characters will initially be empty
+
+
+---
+
+### Exercise 3: Open Trivia Database API
```

Comment:
Probably we will take time for previous exercises and this will become a homework.

Reply:
I was thinking to let people choose. Maybe not everyone likes the Rick and Morty one and they can start directly with this